src/Agent/NewRelic/Agent/Extensions/NewRelic.Agent.Extensions/Helpers/SearchRequestWrapperBase.cs

```diff
@@ -0,0 +1,274 @@
+            segment.GetExperimentalApi().SetSegmentData(datastoreSegmentData).MakeLeaf();
+
+            return segment;
+        }
+
+        protected void TryProcessResponse(IAgent agent, ITransaction transaction, object response, ISegment segment)
+        {
+            try
+            {
+                if (response == null || segment == null)
+                {
+                    return;
+                }
+                var apiCallDetails = GetApiCallDetailsFromResponse(response);
+                var uri = GetUriFromApiCallDetails(apiCallDetails);
+                SetUriOnDatastoreSegment(segment, uri);
+                ReportError(transaction, apiCallDetails);
+
+                segment.End();
+            }
+            catch (Exception ex)
+            {
+                agent.HandleWrapperException(ex);
+            }
+        }
+
+        private void ReportError(ITransaction transaction, object apiCallDetails)
+        {
+            var exceptionGetter = _exceptionGetter ??= GetExceptionGetterFromApiCallDetails(apiCallDetails);
+            var ex = exceptionGetter(apiCallDetails);
+
+            if ((ex != null) && (ex is Exception exception))
+            {
+                transaction.NoticeError(exception);
+                return;
+            }
+
+            // If an error can be caught by the library before the request is made, it doesn't throw an exception, or
+            // set any kind of error object. The best we can do is check if it was successful, and use the ToString()
+            // override to get a summary of what happened
+            var successGetter = _successGetter ??= GetSuccessGetterFromApiCallDetails(apiCallDetails);
+            var success = successGetter(apiCallDetails);
+
+            if (!success)
+            {
+                transaction.NoticeError(new Exception(apiCallDetails.ToString()));
+            }
+
+        }
+
+        private object GetApiCallDetailsFromResponse(object response)
+        {
+            var ApiCallDetailsGetter = _apiCallDetailsGetter ??= GetApiCallDetailsGetterFromResponse(response);
+            var apiCallDetails = ApiCallDetailsGetter.Invoke(response);
+            return apiCallDetails;
+        }
+
+        private Uri GetUriFromApiCallDetails(object apiCallDetails)
+        {
+            var UriGetter = _uriGetter ??= GetUriGetterFromApiCallDetails(apiCallDetails);
+            var uri = UriGetter.Invoke(apiCallDetails);
+
+            return uri;
+        }
+
+        private static Func<object, object> GetApiCallDetailsGetterFromResponse(object response)
+        {
+            var typeOfResponse = response.GetType();
+            var responseAssemblyName = typeOfResponse.Assembly.FullName;
+            var apiCallDetailsPropertyName =
+                responseAssemblyName.StartsWith("OpenSearch.Net") || responseAssemblyName.StartsWith("Elastic.Clients.Elasticsearch")
```

Comment:
Any way to extract this logic so it's not embedded in the base class?

Reply:
Done!